src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java

```diff
@@ -2033,72 +2007,134 @@ AbstractAsyncSSLConnection getConnection() {
-    }
-
-    private boolean markShutdownRequested() {
-        return markClosedState(SHUTDOWN_REQUESTED);
-    }
-
-    private boolean markHalfClosedLocal() {
-        return markClosedState(HALF_CLOSED_LOCAL);
+    private void sendGoAway(final GoAwayFrame goAway) {
+        // currently we send a GOAWAY just once irrespective of what value the
+        // last stream id was in the GOAWAY frame
+        if (!goAwaySent.compareAndSet(false, true)) {
+            // already sent
+            return;
+        }
+        Log.logTrace("{0} sending GOAWAY {1}", connection, goAway);
+        if (debug.on()) {
+            debug.log("sending GOAWAY " + goAway);
+        }
+        // this merely enqueues the frame
+        sendFrame(goAway);
     }
 
-    private boolean markHalfClosedRemote() {
-        return markClosedState(HALF_CLOSED_REMOTE);
-    }
+    // Responsible for doing all the necessary work for closing a Http2Connection
+    private final class Terminator {
+        // the cause for closing the connection. Must only be set in the
+        // Terminator.terminate(Http2TerminationCause) method.
+        private final AtomicReference<Http2TerminationCause> terminationCause = new AtomicReference<>();
+        // true if it has been decided to terminate the connection due to being idle,
+        // false otherwise. should be accessed only when holding the stateLock
+        private boolean chosenForIdleTermination;
 
-    private boolean markIdleShutdownInitiated() {
-        return markClosedState(IDLE_SHUTDOWN_INITIATED);
-    }
+        private void terminate(final Http2TerminationCause terminationCause) {
+            Objects.requireNonNull(terminationCause, "termination cause cannot be null");
+            // allow to be terminated only once
+            stateLock.lock();
+            try {
+                final boolean success = this.terminationCause.compareAndSet(null, terminationCause);
+                if (!success) {
+                    // already terminated or is being terminated by some other thread
+                    return;
+                }
+                // disable the idle timeout event, since we are now going to terminate the
+                // connection
+                Http2Connection.this.cancelIdleCloseEvent();
+            } finally {
+                stateLock.unlock();
+            }
+            // do the actual termination
+            doTerminate();
+        }
 
-    private boolean markClosedState(int flag) {
-        int state, desired;
-        do {
-            state = desired = closedState;
-            if ((state & flag) == flag) return false;
-            desired = state | flag;
-        } while (!CLOSED_STATE.compareAndSet(this, state, desired));
-        return true;
-    }
+        private void doTerminate() {
+            final Http2TerminationCause tc = terminationCause.get();
+            assert tc != null : "missing termination cause";
```

Comment:
`doTerminate()` can receive the termination cause from `terminate()`, which would render these two lines redundant.

Reply:
Actually, this one was intentional. I wanted to avoid introducing a `doTerminate(cause)` (even if `private`) which could then be invoked from arbitrary places with different causes.

In doTerminate() I always want it to use the cause that's already set and setting of that cause must always happen in `terminate(cause)`. terminate(cause) itself is allowed to be called from numerous places and it has the necessary code to synchronize in a way that allows setting the cause just once and then calling doTerminate() just once.